gprofiler/py_package_version.py

```diff
@@ -0,0 +1,160 @@
+"""Package name and version for Python files.
+
+Some of the functions in this module are implemented based on similar functions
+in pip: _get_metadata, _convert_legacy_entry, _files_from_record,
+_files_from_legacy, _get_package_name.
+"""
+import csv
+import email
+import os
+import pathlib
+from typing import Iterator, List, Optional, Tuple
+
+import pkg_resources
+
+# A little monkey patch to prevent pkg_resources from converting "/proc/{pid}/root/" to "/"
+pkg_resources._normalize_cached = lambda path: path
+
+
+__all__ = ["get_versions"]
+
+
+def _convert_to_proc_root_path(path: str, pid: int) -> str:
+    assert path.startswith("/")
+    return os.path.join(f"/proc/{pid}/root", path)
+
+
+def _get_packages_dir(file_path: str) -> Optional[str]:
+    if not file_path.startswith("/"):
+        return None
+
+    idx = file_path.rfind("-packages/")
+    if idx == -1 or (not file_path[:idx].endswith("site") and not file_path[:idx].endswit("dist")):
+        return None
+
+    return file_path[:idx] + "-packages/"
+
+
+def _get_metadata(dist: pkg_resources.Distribution) -> email.message.Message:
+    metadata_name = "METADATA"
+    if isinstance(dist, pkg_resources.DistInfoDistribution) and dist.has_metadata(metadata_name):
+        metadata = dist.get_metadata(metadata_name)
+    elif dist.has_metadata("PKG-INFO"):
+        metadata_name = "PKG-INFO"
+        metadata = dist.get_metadata(metadata_name)
+    else:
+        metadata = None
+
+    if metadata is None:
+        return {}
+
+    feed_parser = email.parser.FeedParser()
+    feed_parser.feed(metadata)
+    return feed_parser.close()
+
+
+def _convert_legacy_entry(entry: Tuple[str, ...], info: Tuple[str, ...]) -> str:
+    """Convert a legacy installed-files.txt path into modern RECORD path.
+
+    The legacy format stores paths relative to the info directory, while the
+    modern format stores paths relative to the package root, e.g. the
+    site-packages directory.
+
+    :param entry: Path parts of the installed-files.txt entry.
+    :param info: Path parts of the egg-info directory relative to package root.
+    :returns: The converted entry.
+
+    For best compatibility with symlinks, this does not use ``abspath()`` or
+    ``Path.resolve()``, but tries to work with path parts:
+
+    1. While ``entry`` starts with ``..``, remove the equal amounts of parts
+       from ``info``; if ``info`` is empty, start appending ``..`` instead.
+    2. Join the two directly.
+    """
+    while entry and entry[0] == "..":
+        if not info or info[-1] == "..":
+            info += ("..",)
+        else:
+            info = info[:-1]
+        entry = entry[1:]
+    return str(pathlib.Path(*info, *entry))
+
+
+def _files_from_record(dist: pkg_resources.Distribution) -> Optional[Iterator[str]]:
+    try:
+        text = dist.get_metadata("RECORD")
+    except:
```

Comment:
There are several unrelated exceptions that can be caught here. Since the operation of this module isn't crucial for gProfiler an error here is always non-critical and should only fail the operation of figuring the version, regardless of the exception type.
Maybe I should log unpredicted exceptions though.

Reply:
`SystemExit` and possibly others, I'm not sure. What you can do is `except Exception` which should not catch those like keyboard interrupt. See https://docs.python.org/3/library/exceptions.html#exception-hierarchy.

For better code readability, I'd suggest:
* At each function / possibly-failing-operation,you catch the specific exceptions it can raise (and place the `try except` on the smallest possible code area)
* The module entrypoint itself should be wrapped with `except Exception` in no way it interferes with gProfiler